File: oldapp/src/main/java/com/riprog/launcher/HomeItem.java
package com.riprog.launcher;

public class HomeItem {
    public enum Type { APP, WIDGET, CLOCK }

    public Type type;
    public String packageName;
    public String className;
    public float row;
    public float col;
    public int spanX = 1;
    public int spanY = 1;
    public int page;
    public int widgetId = -1;

    // Advanced Freeform Transformations
    public float rotation = 0f;
    public float scale = 1.0f;
    public float tiltX = 0f;
    public float tiltY = 0f;

    public HomeItem() {}

    public static HomeItem createApp(String packageName, String className, float col, float row, int page) {
        HomeItem item = new HomeItem();
        item.type = Type.APP;
        item.packageName = packageName;
        item.className = className;
        item.col = col;
        item.row = row;
        item.spanX = 1;
        item.spanY = 1;
        item.page = page;
        return item;
    }

    public static HomeItem createWidget(int widgetId, float col, float row, int spanX, int spanY, int page) {
        HomeItem item = new HomeItem();
        item.type = Type.WIDGET;
        item.widgetId = widgetId;
        item.col = col;
        item.row = row;
        item.spanX = spanX;
        item.spanY = spanY;
        item.page = page;
        return item;
    }

    public static HomeItem createClock(float col, float row, int spanX, int spanY, int page) {
        HomeItem item = new HomeItem();
        item.type = Type.CLOCK;
        item.col = col;
        item.row = row;
        item.spanX = spanX;
        item.spanY = spanY;
        item.page = page;
        return item;
    }
}

File: app/src/main/java/com/riprog/launcher/data/model/HomeItem.kt
package com.riprog.launcher.data.model

class HomeItem {
    enum class Type { APP, WIDGET, CLOCK, FOLDER }


    var type: Type? = null

    var packageName: String? = null

    var className: String? = null

    var folderName: String? = null

    var folderItems: MutableList<HomeItem> = mutableListOf()

    var row: Float = 0f

    var col: Float = 0f

    var spanX: Float = 1f

    var spanY: Float = 1f

    var page: Int = 0

    var originalRow: Float = 0f
    var originalCol: Float = 0f
    var originalSpanX: Float = 1f
    var originalSpanY: Float = 1f
    var originalPage: Int = 0

    var widgetId: Int = -1


    var rotation: Float = 0f

    var scale: Float = 1.0f

    var tiltX: Float = 0f

    var tiltY: Float = 0f

    var isLayoutLocked: Boolean = false

    companion object {

        fun createApp(packageName: String, className: String, col: Float, row: Float, page: Int): HomeItem {
            val item = HomeItem()
            item.type = Type.APP
            item.packageName = packageName
            item.className = className
            item.col = col
            item.row = row
            item.spanX = 1f
            item.spanY = 1f
            item.page = page
            item.originalCol = col
            item.originalRow = row
            item.originalSpanX = 1f
            item.originalSpanY = 1f
            item.originalPage = page
            return item
        }


        fun createWidget(widgetId: Int, col: Float, row: Float, spanX: Int, spanY: Int, page: Int): HomeItem {
            val item = HomeItem()
            item.type = Type.WIDGET
            item.widgetId = widgetId
            item.col = col
            item.row = row
            item.spanX = spanX.toFloat()
            item.spanY = spanY.toFloat()
            item.page = page
            item.originalCol = col
            item.originalRow = row
            item.originalSpanX = spanX.toFloat()
            item.originalSpanY = spanY.toFloat()
            item.originalPage = page
            return item
        }


        fun createClock(col: Float, row: Float, spanX: Int, spanY: Int, page: Int): HomeItem {
            val item = HomeItem()
            item.type = Type.CLOCK
            item.col = col
            item.row = row
            item.spanX = spanX.toFloat()
            item.spanY = spanY.toFloat()
            item.page = page
            item.originalCol = col
            item.originalRow = row
            item.originalSpanX = spanX.toFloat()
            item.originalSpanY = spanY.toFloat()
            item.originalPage = page
            return item
        }


        fun createFolder(name: String, col: Float, row: Float, page: Int): HomeItem {
            val item = HomeItem()
            item.type = Type.FOLDER
            item.folderName = name
            item.col = col
            item.row = row
            item.spanX = 1f
            item.spanY = 1f
            item.page = page
            item.originalCol = col
            item.originalRow = row
            item.originalSpanX = 1f
            item.originalSpanY = 1f
            item.originalPage = page
            return item
        }
    }
}
